File: source python/Organisms/Animals/Animal.py
from Application.Organisms.Organism import *
from Application.Requirements import NoRequirements

class Animal(Organism):

    def __init__(self, strength, initiative, position, world, **kwargs):
        super().__init__(strength, initiative, position, world, **kwargs)

    def Act(self):
        self._MoveInNeighbourhood(NoRequirements())

    def Collide(self, organism):
        return ColisionResult.FIGHT

    def RunCollision(self, organism):   
        if organism is None:
            return True

        if self._Copulate(organism):
            return False

        myCollisionResult = self.Collide(organism)
        organismCollisionResult = organism.Collide(self)

        if myCollisionResult == ColisionResult.ESCAPE:
            print(str(self) + " runs from " + str(organism))
            return False    
        if myCollisionResult == ColisionResult.STOP or organismCollisionResult == ColisionResult.STOP:
            print("Fight beetwen " + str(self) + " and " + str(organism) + " has been stopped")
            return False
        if organismCollisionResult == ColisionResult.ESCAPE:
            print(str(organism) + " runs from " + str(self))
            return True
        
        return self.__Fight(organism)

    def _GoTo(self, position):
        self.GetWorld().MoveAnimal(self, position)

    def _Move(self, vector):
        self._GoTo(self.GetPosition() + vector)

    def _MoveInNeighbourhood(self, requirements):
        newPosition = self.GetWorld().GetRandomPlaceInNeighbourhood(self.GetPosition(), requirements)
        if newPosition is None:
            return False

        self._GoTo(newPosition)
        return True

    def _Copulate(self, organism):
        if type(self) is not type(organism):
            return False

        position = self.GetWorld().GetRandomPlaceInNeighbourhood(self.GetPosition(), NoRequirements())
        if position is None:
            return True
            
        self.GetWorld().CreateOrganism(position, self)
        return True

    def __Fight(self, organism):
        if isinstance(organism, Animal) and self.GetStrength() < organism.GetStrength():
            print(str(organism) + " defeats " + str(self))
            self.Die()
            return False

        print(str(self) + " defeats/eats " + str(organism))
        organism.Die()
        return True

    

File: source java/organism/animals/Animal.java
package organism.animals;

import organism.CollisionResult;
import organism.Organism;
import Interfaces.Requirements;
import utilities.position.Position;
import utilities.readyclasses.NoRequirements;
import utilities.savingsystem.Loader;
import virtualworld.World;

public abstract class Animal extends Organism
{
    
    public Animal(int strength, int initiative, Position position, World world)
    {
        super(strength, initiative, position, world);
    }
    
    public Animal(Loader loader, World world, Position position)
    {
        super(loader, world, position);
    }
    
    @Override
    public void Act()
    {
        MoveInNeighbourhood(new NoRequirements());
    }
    
    @Override
    public CollisionResult Collide(Organism organism)
    {
        return CollisionResult.FIGHT;
    }
    
    public boolean RunCollision(Organism organism)
    {
        if(organism == null)
            return true;
        
        if(Copulate(organism))
            return false;
        
        CollisionResult myCollisionResult = Collide(organism);
        CollisionResult organismCollisionResult = organism.Collide(this);
        
        if(myCollisionResult == CollisionResult.ESCAPE)
        {
            GetWorld().AddToLogs(toString() + " runs away from " + organism.toString());
            return false;
        }
        if(myCollisionResult == CollisionResult.STOP || organismCollisionResult == CollisionResult.STOP)
        {
            GetWorld().AddToLogs("Fight beetwen " + organism.toString() + " and " + toString() + " has been stopped");
            return false;
        }
        if(organismCollisionResult == CollisionResult.ESCAPE)
        {
            GetWorld().AddToLogs(organism.toString() + " runs away from " + toString());
            return true;
        }
        
     
        return Fight(organism);
    }
    
    boolean Fight(Organism organism)
    {
        if(organism instanceof Animal && this.GetStrength() < organism.GetStrength())
        {
            GetWorld().AddToLogs(organism.toString() + " defeats " + toString());
            Die();
            return false;
        }
        
        organism.Die();
        GetWorld().AddToLogs(toString() + " defeats/eats " + organism.toString());
        return true;
    }
    
    protected boolean Copulate(Organism organism)
    {
        if(organism.GetEnum() != GetEnum())
            return false;
        
        class Req implements Requirements
        {
            @Override
            public boolean IsOk(Organism organism)
            {
                return organism == null;
            }
        }
       
        Position position = GetWorld().GetRandomPlaceInNeighbourhood(GetPosition(), new Req());
        if(position == null)
            return true;
        
        GetWorld().CreateOrganism(GetEnum(), position);
        return true;
    }
    
    protected boolean MoveInNeighbourhood(Requirements req)
    {
        Position newPosition = GetWorld().GetRandomPlaceInNeighbourhood(GetPosition(), req);
        if(newPosition == null)
            return false;
        GoTo(newPosition);
        return true;
    }
    
    void GoTo(Position position)
    {
        GetWorld().MoveAnimal(this, position);
    }
}
